### ISSUE TITLE
S5 6.Интеграция AI-модерации контента и проверки на плагиат

### ISSUE THREAD

**Issue by AFZhuk63** [AFZhuk63/Chatty_green#78]
6.1. Цели
- Автоматическая цензура: Блокировка запрещенного контента (маты, порнография, наркотики, оружие).
- Проверка на плагиат: Выявление копирования чужих материалов.
- Юридическая безопасность: Защита от DMCA-жалоб и нарушений авторских прав.

Техническая реализация
🔞 AI-цензура контента

Инструменты:
- Google Cloud Vision API (для изображений/видео)
- Yandex Content Moderator или AWS Rekognition (альтернативы)
- NLP-библиотеки (для текста): transformers, natasha (для русского языка)

Код для текстовой модерации:
# Пример с библиотекой transformers (Hugging Face)

from transformers import pipeline

moderator = pipeline("text-classification", model="cointegrated/rubert-tiny-toxicity")

def is_toxic(text: str, threshold=0.9) -> bool:
    result = moderator(text)[0]
    return result["label"] == "toxic" and result["score"] > threshold

Для изображений:

# Google Cloud Vision (аналог для AWS/Azure)
from google.cloud import vision

client = vision.ImageAnnotatorClient()
response = client.safe_search_detection(image=image)
print(response.safe_search_annotation.adult)  # Порнография?

Проверка на плагиат
Сервисы:

- Copyscape API (платный)
- Originality.ai (для глубокого анализа)
- Самописное решение (если бюджет ограничен):

# Базовая проверка схожести текста (TF-IDF + cosine similarity)
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

def check_plagiarism(text1, text2, threshold=0.8):
    vectorizer = TfidfVectorizer()
    tfidf = vectorizer.fit_transform([text1, text2])
    similarity = cosine_similarity(tfidf[0], tfidf[1])[0][0]
    return similarity > threshold

# Модель для хранения результатов модерации

class ContentModerationLog(models.Model):
    content = models.ForeignKey(Post, on_delete=models.CASCADE)
    is_approved = models.BooleanField(default=False)
    reason = models.TextField()  # "Обнаружены маты", "Плагиат" и т.д.
    checked_at = models.DateTimeField(auto_now_add=True)

6.3. Критерии выполнения

- AI-модератор проверяет тексты/изображения при загрузке.
- Система блокирует контент с toxicity-score > 90%.
- Проверка на плагиат работает для текстов (>80% схожести = отклонение).
- Интегрирована форма для DMCA-жалоб.

6.4 Дополнительные задачи
Кастомные правила: Возможность добавлять стоп-слова через админку.
Модерация комментариев: Автоматическое скрытие токсичных сообщений.
Апелляции: Если контент заблокирован ошибочно, автор может подать апелляцию.

Пример кастомных правил в админке:

class BannedWord(models.Model):
    word = models.CharField(max_length=100)
    reason = models.CharField(max_length=200)  # "Матерное слово", "Пропаганда" и т.д.
